proto-lens/src/Data/ProtoLens/Encoding/Bytes.hs

```diff
@@ -57,11 +58,27 @@ anyBits = loop 0 0
             b <- anyWord8
             loop (w .|. shiftL (fromIntegral b) n) (n+8)
 
-putVarInt :: Word64 -> Builder
-putVarInt n
-    | n < 128 = Builder.word8 (fromIntegral n)
-    | otherwise = Builder.word8 (fromIntegral $ n .&. 127 .|. 128)
-                      <> putVarInt (n `shiftR` 7)
+-- | A wrapper around a 'Builder' that keeps track of its size.
+data SizedBuilder = SizedBuilder
+  { _size :: Word64 -- Purposefully lazy
+  , _builder :: !Builder
+  }
+
+instance Semigroup SizedBuilder where
+  (SizedBuilder s1 b1) <> (SizedBuilder s2 b2) = SizedBuilder (s1 + s2) (b1 <> b2)
+
+instance Monoid SizedBuilder where
+  mempty = SizedBuilder 0 mempty
+  mappend = (<>)
+
+putVarInt :: Word64 -> SizedBuilder
+putVarInt x = SizedBuilder sz $ putVarInt' x
+  where
+    sz = fromIntegral $ 1 + ((64 - countLeadingZeros x) `quot` 7)
+    putVarInt' n
+        | n < 128 = Builder.word8 (fromIntegral n)
+        | otherwise = Builder.word8 (fromIntegral $ n .&. 127 .|. 128)
+                      <> putVarInt' (n `shiftR` 7)
```

Comment:
@judah @shlevy note that you could use http://hackage.haskell.org/package/bytestring-0.10.8.2/docs/Data-ByteString-Builder-Prim.html to avoid the recursion completely and perform the check whether the buffer is large enough in a single step. See the `encodeCharUtf8HtmlEscaped` example for inspiration.